File: AS_VALs/demos/python/demo_nonlinear_motion_blur.py
#!/usr/bin/env python3
"""Differentiable nonlinear (spatially-variant) motion blur — AS_VALs demo.

Reproduces the MATLAB experiment using `motion_blur` (same API as
`Noise_Layer.ScreenShootingMB`).

Run::

    python AS_VALs/demos/python/demo_nonlinear_motion_blur.py
"""

from __future__ import annotations

import math
import random
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import torch
from PIL import Image

THIS_DIR = Path(__file__).resolve().parent
AS_VALS = THIS_DIR.parents[1]
REPO_ROOT = THIS_DIR.parents[2]
CORE_DIR = AS_VALS / "python_motion_blur"
TEXT_IMAGES = AS_VALS / "outputs" / "text_images"
OUT_DIR = AS_VALS / "outputs" / "demo_python"

if str(CORE_DIR) not in sys.path:
    sys.path.insert(0, str(CORE_DIR))

from motion_blur import (  # noqa: E402
    nonlinear_motion_blur,
    sample_pose_pair,
    sharp_view,
)


# ---------------------------------------------------------------------------
# I/O + visualization
# ---------------------------------------------------------------------------


def load_image_tensor(path: Path, device: torch.device) -> torch.Tensor:
    img = Image.open(path).convert("RGB")
    arr = np.asarray(img, dtype=np.float32) / 255.0  # HWC RGB
    ten = torch.from_numpy(arr).permute(2, 0, 1).unsqueeze(0).to(device)
    return ten


def to_numpy_img(t: torch.Tensor) -> np.ndarray:
    """(B,C,H,W) or (C,H,W) → HWC uint-display float in [0,1]."""
    x = t.detach()
    if x.ndim == 4:
        x = x[0]
    x = x.permute(1, 2, 0).cpu().numpy()
    return np.clip(x, 0.0, 1.0)


def plot_results(
    image: torch.Tensor,
    sharp: torch.Tensor,
    blurred: torch.Tensor,
    aux: dict[str, torch.Tensor],
    cx0: float,
    cy0: float,
    theta0: float,
    cx1: float,
    cy1: float,
    theta1: float,
    img_name: str,
    traj_grid: int = 16,
    quiver_step: int = 24,
) -> plt.Figure:
    """Four-panel figure matching the MATLAB experiment."""
    h, w = image.shape[-2], image.shape[-1]
    half = (sharp.shape[-1] - 1) / 2.0
    src = to_numpy_img(image)
    mag = aux["mag"].detach().cpu().numpy()
    x0 = aux["x0"].detach().cpu().numpy()
    y0 = aux["y0"].detach().cpu().numpy()
    x1 = aux["x1"].detach().cpu().numpy()
    y1 = aux["y1"].detach().cpu().numpy()
    u_grid = aux["u_grid"].detach().cpu().numpy()
    v_grid = aux["v_grid"].detach().cpu().numpy()

    fig = plt.figure(figsize=(14, 10), facecolor="white", constrained_layout=True)

    # 1) original
    ax1 = fig.add_subplot(2, 3, 1)
    ax1.imshow(src)
    ax1.set_title(f"Original {img_name}\n(600×600)")
    ax1.axis("off")

    # 2) sharp vs blur
    ax2 = fig.add_subplot(2, 3, 2)
    ax2.imshow(to_numpy_img(sharp))
    ax2.set_title("Sharp start crop (512)")
    ax2.axis("off")

    ax2b = fig.add_subplot(2, 3, 3)
    ax2b.imshow(to_numpy_img(blurred))
    ax2b.set_title("Nonlinear motion blur")
    ax2b.axis("off")

    # 3) magnitude heatmap
    ax3 = fig.add_subplot(2, 3, 4)
    im = ax3.imshow(mag, cmap="viridis")
    fig.colorbar(im, ax=ax3, fraction=0.046, pad=0.04)
    ax3.set_title(f"Trajectory length |p1−p0|\n[{mag.min():.2f}, {mag.max():.2f}] px")
    ax3.set_xlabel("u")
    ax3.set_ylabel("v")

    # 4) sampled paths on original
    ax4 = fig.add_subplot(2, 3, 5)
    ax4.imshow(src)
    uu = u_grid[::traj_grid, ::traj_grid]
    vv = v_grid[::traj_grid, ::traj_grid]
    t_fine = np.linspace(0.0, 1.0, 12)
    for ui, vi in zip(uu.ravel(), vv.ravel()):
        xs, ys = [], []
        for t in t_fine:
            cx = (1 - t) * cx0 + t * cx1
            cy = (1 - t) * cy0 + t * cy1
            th = (1 - t) * theta0 + t * theta1
            ct, st = math.cos(th), math.sin(th)
            xs.append(ct * ui - st * vi + cx)
            ys.append(st * ui + ct * vi + cy)
        ax4.plot(xs, ys, "-", color=(0.1, 0.45, 0.95, 0.85), linewidth=0.8)
        ax4.plot(xs[0], ys[0], "g.", markersize=3)
        ax4.plot(xs[-1], ys[-1], "r.", markersize=3)

    xq = x0[::quiver_step, ::quiver_step]
    yq = y0[::quiver_step, ::quiver_step]
    dxq = x1[::quiver_step, ::quiver_step] - xq
    dyq = y1[::quiver_step, ::quiver_step] - yq
    ax4.quiver(
        xq, yq, dxq, dyq, color="gold", angles="xy", scale_units="xy", scale=1, width=0.002
    )
    ax4.set_xlim(0.5, w + 0.5)
    ax4.set_ylim(h + 0.5, 0.5)
    ax4.set_title("Trajectories (green=start, red=end)")
    ax4.set_aspect("equal")

    # 5) start / end viewfinder
    ax5 = fig.add_subplot(2, 3, 6)
    ax5.imshow(src)
    corners = np.array(
        [[-half, -half], [half, -half], [half, half], [-half, half], [-half, -half]],
        dtype=np.float64,
    )

    def poly_for(cx, cy, th):
        ct, st = math.cos(th), math.sin(th)
        xs = ct * corners[:, 0] - st * corners[:, 1] + cx
        ys = st * corners[:, 0] + ct * corners[:, 1] + cy
        return xs, ys

    xs0, ys0 = poly_for(cx0, cy0, theta0)
    xs1, ys1 = poly_for(cx1, cy1, theta1)
    ax5.plot(xs0, ys0, "g-", linewidth=2.0, label=r"start $\theta=0$")
    ax5.plot(
        xs1,
        ys1,
        "r-",
        linewidth=2.0,
        label=rf"end $\theta={math.degrees(theta1):.1f}^\circ$",
    )
    ax5.plot(cx0, cy0, "g+", markersize=12, markeredgewidth=1.5)
    ax5.plot(cx1, cy1, "rx", markersize=10, markeredgewidth=1.5)
    t_c = np.linspace(0, 1, 40)
    ax5.plot(
        (1 - t_c) * cx0 + t_c * cx1,
        (1 - t_c) * cy0 + t_c * cy1,
        "c--",
        linewidth=1.2,
        label="center path",
    )
    ax5.legend(loc="lower center", fontsize=8, ncol=2)
    ax5.set_title("Start / end viewfinder")
    ax5.set_xlim(0.5, w + 0.5)
    ax5.set_ylim(h + 0.5, 0.5)
    ax5.set_aspect("equal")

    fig.suptitle("Differentiable nonlinear motion blur (PyTorch grid_sample)", fontsize=13)
    return fig


def check_differentiable(blurred: torch.Tensor, image: torch.Tensor) -> None:
    loss = blurred.mean()
    loss.backward()
    assert image.grad is not None, "image.grad is None — blur is not differentiable"
    g = image.grad
    print(
        f"[grad check] OK  |grad| mean={g.abs().mean().item():.6e}  "
        f"max={g.abs().max().item():.6e}  nonzero={int((g != 0).sum())}/{g.numel()}"
    )


def main() -> None:
    img_dir = TEXT_IMAGES
    paths = sorted(img_dir.glob("*.png"))
    if not paths:
        raise FileNotFoundError(f"No PNGs under {img_dir}")

    img_path = random.choice(paths)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    view_size = 512
    n_samples = 16
    path_power = 2.0
    max_trans = 28.0
    max_rot_deg = 10.0

    image = load_image_tensor(img_path, device)
    _, _, h, w = image.shape
    if (h, w) != (600, 600):
        raise ValueError(f"expected 600x600 AS_VALs, got {h}x{w}")

    cx0, cy0, theta0, cx1, cy1, theta1, n_reject = sample_pose_pair(
        width=w,
        height=h,
        view_size=view_size,
        max_trans=max_trans,
        max_rot_deg=max_rot_deg,
        n_check=n_samples,
    )

    print(f"image: {img_path.name}")
    print(f"device: {device}")
    print(f"path_power={path_power} (λ=t^p, uniform shutter)")
    print(f"reject_count: {n_reject}")
    print(f"start: cx={cx0:.2f} cy={cy0:.2f} theta={math.degrees(theta0):.2f} deg")
    print(f"end:   cx={cx1:.2f} cy={cy1:.2f} theta={math.degrees(theta1):.2f} deg")

    image = image.requires_grad_(True)
    blurred, aux = nonlinear_motion_blur(
        image,
        cx0,
        cy0,
        theta0,
        cx1,
        cy1,
        theta1,
        view_size=view_size,
        n_samples=n_samples,
        path_power=path_power,
    )
    sharp = sharp_view(image, cx0, cy0, theta0, view_size=view_size)

    check_differentiable(blurred, image)

    out_dir = OUT_DIR
    out_dir.mkdir(parents=True, exist_ok=True)

    # Visualization (detach inside helpers); save + show
    with torch.no_grad():
        fig = plot_results(
            image,
            sharp,
            blurred,
            aux,
            cx0,
            cy0,
            theta0,
            cx1,
            cy1,
            theta1,
            img_path.name,
        )
        save_path = out_dir / f"demo_{img_path.stem}.png"
        fig.savefig(save_path, dpi=150, bbox_inches="tight")
        print(f"saved figure: {save_path}")
        plt.show()


if __name__ == "__main__":
    main()

File: AS_VALs/demos/matlab/demo_nonlinear_motion_blur.m
%% demo_nonlinear_motion_blur
% 空变非线性运动模糊实验（AS_VALs 600x600 → 512x512 取景窗）
%
% 思想：
%   - 起点位姿 g0：随机平移，旋转 θ0 = 0
%   - 终点位姿 g1：随机平移 + 随机旋转 θ
%   - 对曝光时间 t∈[0,1] 做刚体位姿插值 g(t)
%   - 每个输出像素沿自身轨迹双线性采样，高斯加权积分
%
% 输出四张 figure，并保存到 out/：
%   1) 原图 (600x600)
%   2) 结果图 (512 模糊)
%   3) 轨迹图 (位移场 + 抽样轨迹)
%   4) 起点/终点取景窗

clear; close all; clc;
rng(42);  % 固定种子，便于复现；改 [] 则每次随机

%% -------------------- 路径与输入 --------------------
thisDir = fileparts(mfilename('fullpath'));
imgDir  = fullfile(thisDir, '..', '..', 'outputs', 'text_images');
outDir  = fullfile(thisDir, '..', '..', 'outputs', 'demo_matlab');
if ~exist(outDir, 'dir'); mkdir(outDir); end

% 选一张 AS_VALs 图；也可改成具体文件名
imgName = '00010.png';  % manifest 里是 "new book"，文字较清晰
imgPath = fullfile(imgDir, imgName);
if ~isfile(imgPath)
    error('找不到图像: %s\n请确认 AS_VALs/outputs/text_images 下已有 PNG。', imgPath);
end

I = im2double(imread(imgPath));
if size(I, 3) == 1
    I = repmat(I, [1 1 3]);
end
[H, W, C] = size(I);
assert(H == 600 && W == 600, '期望 AS_VALs 为 600x600，实际为 %dx%d', H, W);

%% -------------------- 参数 --------------------
viewSize = 512;          % 取景窗边长
nSamples = 16;           % 轨迹时间采样数（越大越平滑、越慢）
pathPower = 2;           % 进度 λ=t^pathPower；2=匀加速(v∝t)，1=匀速
maxTrans = 28;           % 终点相对起点的最大平移（像素）
maxRotDeg = 10;          % 终点最大旋转角（度）
maxPoseTries = 500;      % 越界则重采样的最大次数
trajGrid = 16;           % 轨迹图上每隔多少像素画一条示意轨迹
quiverStep = 24;         % 位移场箭头稀疏步长

half = (viewSize - 1) / 2;   % 局部坐标相对窗中心：[-half, half]
% 取景窗四角（局部坐标），用于越界检查
cornerU = [-half,  half, half, -half];
cornerV = [-half, -half, half,  half];

%% -------------------- 随机起点 / 终点位姿（越界则重采） --------------------
% 全程（起点、终点、以及插值中间位姿）取景窗四角都必须落在 [1,W]×[1,H] 内
theta0 = 0;
ok = false;
nReject = 0;
for tryId = 1:maxPoseTries
    % 起点：θ=0，先保证轴对齐窗可放入图内
    cx0 = half + 1 + rand() * (W - viewSize);
    cy0 = half + 1 + rand() * (H - viewSize);

    % 终点：起点附近平移 + 旋转
    cx1 = cx0 + (2 * rand() - 1) * maxTrans;
    cy1 = cy0 + (2 * rand() - 1) * maxTrans;
    theta1 = deg2rad((2 * rand() - 1) * maxRotDeg);

    if pose_path_in_bounds(cornerU, cornerV, cx0, cy0, theta0, ...
            cx1, cy1, theta1, W, H, nSamples)
        ok = true;
        break;
    end
    nReject = nReject + 1;
end
if ~ok
    error(['连续 %d 次随机位姿均越界。请减小 maxTrans / maxRotDeg，' ...
           '或增大 maxPoseTries。'], maxPoseTries);
end

fprintf('图像: %s\n', imgPath);
fprintf('越界重采次数: %d\n', nReject);
fprintf('起点: cx=%.2f cy=%.2f theta=%.2f deg\n', cx0, cy0, rad2deg(theta0));
fprintf('终点: cx=%.2f cy=%.2f theta=%.2f deg\n', cx1, cy1, rad2deg(theta1));

%% -------------------- 局部网格（输出 512x512） --------------------
[uGrid, vGrid] = meshgrid(linspace(-half, half, viewSize), ...
                          linspace(-half, half, viewSize));

%% -------------------- 刚体插值（匀加速进度）+ 等权曝光 --------------------
tList = linspace(0, 1, nSamples);
lamList = tList .^ pathPower;   % λ(t)=t^p；p=2 → 静止起步匀加速
wList = ones(1, nSamples) / nSamples;

acc = zeros(viewSize, viewSize, C);
[x0map, y0map] = pose_map(uGrid, vGrid, cx0, cy0, theta0);
[x1map, y1map] = pose_map(uGrid, vGrid, cx1, cy1, theta1);

for k = 1:nSamples
    lam = lamList(k);
    cx = (1 - lam) * cx0 + lam * cx1;
    cy = (1 - lam) * cy0 + lam * cy1;
    th = (1 - lam) * theta0 + lam * theta1;
    [xs, ys] = pose_map(uGrid, vGrid, cx, cy, th);

    frame = zeros(viewSize, viewSize, C);
    for ch = 1:C
        % 位姿已保证不越界；extrapolate 值仅作兜底
        frame(:, :, ch) = interp2(I(:, :, ch), xs, ys, 'linear', 1);
    end
    acc = acc + wList(k) * frame;
end
Iblur = acc;

% 仅起点清晰裁切（对照）
Isharp = zeros(viewSize, viewSize, C);
for ch = 1:C
    Isharp(:, :, ch) = interp2(I(:, :, ch), x0map, y0map, 'linear', 1);
end

% 运动场：终点相对起点的位移（输出像素坐标系下）
dx = x1map - x0map;
dy = y1map - y0map;
mag = hypot(dx, dy);

%% -------------------- Figure 1: 原图 --------------------
fig1 = figure('Name', '1-原图', 'Color', 'w', 'Position', [80 80 640 680]);
imshow(I);
title(sprintf('原图 AS\\_VALs %s (600\\times600)', imgName), 'Interpreter', 'tex');
export_fig(fig1, fullfile(outDir, '01_original.png'));

%% -------------------- Figure 2: 结果图 --------------------
fig2 = figure('Name', '2-结果图', 'Color', 'w', 'Position', [100 100 980 520]);
subplot(1, 2, 1);
imshow(Isharp);
title('起点清晰取景 (512\times512)');
subplot(1, 2, 2);
imshow(Iblur);
title(sprintf('非线性运动模糊结果 (N=%d, p=%.2f)', nSamples, pathPower));
export_fig(fig2, fullfile(outDir, '02_result.png'));
imwrite(Iblur, fullfile(outDir, '02_result_only.png'));

%% -------------------- Figure 3: 轨迹图 --------------------
fig3 = figure('Name', '3-轨迹图', 'Color', 'w', 'Position', [120 60 1100 520]);

% 左：位移幅度热力图
subplot(1, 2, 1);
imagesc(mag); axis image; colormap(gca, parula); colorbar;
title(sprintf('逐像素轨迹长度 |p_1-p_0|  [%.2f, %.2f] px', min(mag(:)), max(mag(:))));
xlabel('u'); ylabel('v');

% 右：在原图上画抽样轨迹（刚体插值折线）
subplot(1, 2, 2);
imshow(I); hold on;
uu = uGrid(1:trajGrid:end, 1:trajGrid:end);
vv = vGrid(1:trajGrid:end, 1:trajGrid:end);
nShow = numel(uu);
tFine = linspace(0, 1, 12);
for i = 1:nShow
    ui = uu(i); vi = vv(i);
    xs = zeros(size(tFine));
    ys = zeros(size(tFine));
    for k = 1:numel(tFine)
        t = tFine(k);
        cx = (1 - t) * cx0 + t * cx1;
        cy = (1 - t) * cy0 + t * cy1;
        th = (1 - t) * theta0 + t * theta1;
        [xs(k), ys(k)] = pose_map(ui, vi, cx, cy, th);
    end
    plot(xs, ys, '-', 'Color', [0.1 0.45 0.95 0.85], 'LineWidth', 1.0);
    plot(xs(1), ys(1), 'g.', 'MarkerSize', 8);
    plot(xs(end), ys(end), 'r.', 'MarkerSize', 8);
end
% 稀疏 quiver（起点处）
[xq, yq] = pose_map(uGrid(1:quiverStep:end, 1:quiverStep:end), ...
                    vGrid(1:quiverStep:end, 1:quiverStep:end), cx0, cy0, theta0);
dxq = x1map(1:quiverStep:end, 1:quiverStep:end) - xq;
dyq = y1map(1:quiverStep:end, 1:quiverStep:end) - yq;
quiver(xq, yq, dxq, dyq, 0, 'y', 'LineWidth', 0.8);
title('原图上的抽样轨迹 (绿=起点, 红=终点, 黄=位移箭头)');
hold off;
export_fig(fig3, fullfile(outDir, '03_trajectory.png'));

%% -------------------- Figure 4: 起点 / 终点取景窗 --------------------
fig4 = figure('Name', '4-起点终点', 'Color', 'w', 'Position', [140 80 700 720]);
imshow(I); hold on;
corners = [-half -half; half -half; half half; -half half; -half -half];
poly0 = zeros(5, 2);
poly1 = zeros(5, 2);
for i = 1:5
    [poly0(i, 1), poly0(i, 2)] = pose_map(corners(i, 1), corners(i, 2), cx0, cy0, theta0);
    [poly1(i, 1), poly1(i, 2)] = pose_map(corners(i, 1), corners(i, 2), cx1, cy1, theta1);
end
plot(poly0(:, 1), poly0(:, 2), 'g-', 'LineWidth', 2.2);
plot(poly1(:, 1), poly1(:, 2), 'r-', 'LineWidth', 2.2);
plot(cx0, cy0, 'g+', 'MarkerSize', 14, 'LineWidth', 1.5);
plot(cx1, cy1, 'rx', 'MarkerSize', 12, 'LineWidth', 1.5);
% 中心运动轨迹
tFine = linspace(0, 1, 40);
cxs = (1 - tFine) * cx0 + tFine * cx1;
cys = (1 - tFine) * cy0 + tFine * cy1;
plot(cxs, cys, 'c--', 'LineWidth', 1.4);
legend({'起点取景窗 \theta=0', sprintf('终点取景窗 \\theta=%.1f°', rad2deg(theta1)), ...
        '起点中心', '终点中心', '中心轨迹'}, ...
       'Location', 'southoutside', 'NumColumns', 2);
title('起点 / 终点取景窗叠加在原图上');
hold off;
export_fig(fig4, fullfile(outDir, '04_start_end.png'));

%% -------------------- 参数摘要 --------------------
fid = fopen(fullfile(outDir, 'run_info.txt'), 'w');
fprintf(fid, 'image=%s\n', imgPath);
fprintf(fid, 'viewSize=%d\n', viewSize);
fprintf(fid, 'nSamples=%d\n', nSamples);
fprintf(fid, 'pathPower=%.4f\n', pathPower);
fprintf(fid, 'reject_count=%d\n', nReject);
fprintf(fid, 'start: cx=%.4f cy=%.4f theta_deg=%.4f\n', cx0, cy0, rad2deg(theta0));
fprintf(fid, 'end:   cx=%.4f cy=%.4f theta_deg=%.4f\n', cx1, cy1, rad2deg(theta1));
fprintf(fid, 'mag_px: min=%.4f max=%.4f mean=%.4f\n', min(mag(:)), max(mag(:)), mean(mag(:)));
fclose(fid);

fprintf('\n已保存到: %s\n', outDir);
fprintf('  01_original.png\n  02_result.png / 02_result_only.png\n');
fprintf('  03_trajectory.png\n  04_start_end.png\n  run_info.txt\n');

%% ========================================================================
function [X, Y] = pose_map(U, V, cx, cy, theta)
% 局部坐标 (U,V)（相对取景窗中心）→ 原图坐标 (X,Y)
ct = cos(theta);
st = sin(theta);
X = ct .* U - st .* V + cx;
Y = st .* U + ct .* V + cy;
end

function tf = pose_in_bounds(U, V, cx, cy, theta, W, H)
% 取景窗采样点是否全部落在图像内（含边界）
[X, Y] = pose_map(U, V, cx, cy, theta);
tol = 1e-6;
tf = all(X(:) >= 1 - tol & X(:) <= W + tol & ...
         Y(:) >= 1 - tol & Y(:) <= H + tol);
end

function tf = pose_path_in_bounds(U, V, cx0, cy0, theta0, cx1, cy1, theta1, W, H, nCheck)
% 检查起点、终点及刚体插值中间位姿是否都不越界
if nargin < 11 || isempty(nCheck)
    nCheck = 16;
end
tList = linspace(0, 1, nCheck);
tf = true;
for k = 1:numel(tList)
    t = tList(k);
    cx = (1 - t) * cx0 + t * cx1;
    cy = (1 - t) * cy0 + t * cy1;
    th = (1 - t) * theta0 + t * theta1;
    if ~pose_in_bounds(U, V, cx, cy, th, W, H)
        tf = false;
        return;
    end
end
end

function export_fig(figHandle, filePath)
% 优先 exportgraphics；旧版 MATLAB 回退 print
if exist('exportgraphics', 'file') == 2
    exportgraphics(figHandle, filePath, 'Resolution', 150);
else
    set(figHandle, 'PaperPositionMode', 'auto');
    print(figHandle, filePath, '-dpng', '-r150');
end
fprintf('saved: %s\n', filePath);
end
